File: algorithms/searching/traversal/breadth_first_search.py
def breadth_first_search(root):
    result = []
    queue = []

    queue.append(root)

    while queue:
        current_node = queue.pop(0)
        result.append(current_node.value)
        if current_node.left:
            queue.append(current_node.left)
        if current_node.right:
            queue.append(current_node.right)

    return result


def breadth_first_search_recursive(queue, result=[]):
    if not queue:
        return result

    current_node = queue.pop(0)
    result.append(current_node.value)
    if current_node.left:
        queue.append(current_node.left)
    if current_node.right:
        queue.append(current_node.right)

    return breadth_first_search_recursive(queue, result)

File: algorithms/searching/traversal/breadth_first_search.js
function breadthFirstSearch(root) {
    let result = [];
    let queue = [];

    queue.push(root);

    while (queue.length > 0) {
        let currentNode = queue.shift();
        result.push(currentNode.value);
        if (currentNode.left) {
            queue.push(currentNode.left);
        }
        if (currentNode.right) {
            queue.push(currentNode.right);
        }
    }

    return result;
}

function breadthFirstSearchRecursive(queue, result=[]) {
    if (!queue.length > 0) {
        return result;
    }

    let currentNode = queue.shift();
    result.push(currentNode.value);
    if (currentNode.left) {
        queue.push(currentNode.left);
    }
    if (currentNode.right) { 
        queue.push(currentNode.right);
    }

    return breadthFirstSearchRecursive(queue, result);
}